android/src/main/kotlin/dev/msfjarvis/claw/android/ui/LobstersApp.kt

```diff
@@ -68,9 +64,43 @@ fun LobstersApp(
   val postActions = rememberPostActions(urlLauncher, navController, viewModel)
   val currentDestination by currentNavigationDestination(navController)
 
-  val hottestPosts = viewModel.hottestPosts.collectAsLazyPagingItems()
-  val newestPosts = viewModel.newestPosts.collectAsLazyPagingItems()
-  val savedPosts by viewModel.savedPosts.collectAsState(emptyMap())
+  val hottestPostsState by viewModel.hottestPosts.collectAsState(LobstersPagingResult.Loading)
+  val newestPostsState by viewModel.newestPosts.collectAsState(LobstersPagingResult.Loading)
+  var hottestPosts by remember { mutableStateOf(emptyList<LobstersPost>()) }
+  var newestPosts by remember { mutableStateOf(emptyList<LobstersPost>()) }
+  val savedPosts by viewModel.savedPosts.collectAsState(emptyList())
+
+  LaunchedEffect(hottestPostsState) {
+    if (hottestPostsState is LobstersPagingResult.Success) {
+      hottestPosts = (hottestPostsState as LobstersPagingResult.Success).list
+    }
+  }
+
+  LaunchedEffect(hottestListState) {
+    snapshotFlow { hottestListState.firstVisibleItemIndex }
+      .collect { index ->
+        // If only 10 posts are left, load new posts
+        if (hottestPosts.isNotEmpty() && index > hottestPosts.size - 10) {
+          viewModel.loadHottestPosts()
+        }
+      }
+  }
+
+  LaunchedEffect(newestPostsState) {
+    if (newestPostsState is LobstersPagingResult.Success) {
+      newestPosts = (newestPostsState as LobstersPagingResult.Success).list
+    }
+  }
+
+  LaunchedEffect(newestListState) {
+    snapshotFlow { newestListState.firstVisibleItemIndex }
+      .collect { index ->
+        // If only 10 posts are left, load new posts
+        if (newestPosts.isNotEmpty() && index > newestPosts.size - 10) {
+          viewModel.loadNewestPosts()
+        }
+      }
+  }
```

Comment:
I wonder if there should be separate repositories/data sources/use cases/whatever for both the lists' paging related stuff because this is getting kind of ridiculous

Reply:
Yeah, we should definitely clean this up